src/rp2_common/pico_rand/rand.c

```diff
@@ -0,0 +1,293 @@
+    }
+
+    // Return the result.  Note: we don't shift right
+    return fc->result;
+}
+
+
+static uint32_t measure_rosc_freq(void) {
+
+    uint32_t rosc_freq;
+
+    uint32_t snapshot = rosc_hw->ctrl;
+
+    assert(rosc_hw->status & ROSC_STATUS_ENABLED_BITS);
+
+    // speed up rosc, wait until stable
+    rosc_hw->ctrl = (ROSC_CTRL_ENABLE_VALUE_ENABLE   << ROSC_CTRL_ENABLE_LSB) |
+                    (ROSC_CTRL_FREQ_RANGE_VALUE_HIGH << ROSC_CTRL_FREQ_RANGE_LSB);
+    while(!(rosc_hw->status & ROSC_STATUS_STABLE_BITS)) {
+        tight_loop_contents();
+    }
+
+    rosc_freq = frequency_count_rosc();
+
+    // restore rosc (reduces power)
+    rosc_hw->ctrl = snapshot;
+    while(!(rosc_hw->status & ROSC_STATUS_STABLE_BITS)) {
+        tight_loop_contents();
+    }
+
+    return rosc_freq;
+}
+
+
+static void update_rosc_samples(void) {
+
+    static uint32_t previous_sample_time_us = 0;
+
+    assert(rosc_hw->status & ROSC_STATUS_ENABLED_BITS);
+
+    // Ensure that the ROSC random bit is not sampled too quickly,
+    // ROSC may be ticking only a few times a microsecond.
+    // Note: In general (i.e. sporadic) use, very often there will be no delay here.
+    while ((time_us_32() - previous_sample_time_us) < RAND_RAND_BIT_MIN_SAMPLE_US ) {
+       tight_loop_contents();
+    }
+    previous_sample_time_us = time_us_32();
+
+	rosc_samples <<= 1;
+    rosc_samples |= (uint64_t) (rosc_hw->randombit & 1ull);
+}
+
+
+static uint64_t get_new_rand_64(void) {
+
+    if (!rng_initialised) {
+        // Do not provide 'RNs' until the system has been initialised
+        // Note: The first initialisation is quite time consuming at ~80ms,
+        // it is best to use the supplied API function: pico_initialise_rand()
+        // during application initialisation rather than later rely on this hook.
+        pico_initialise_rand(NULL, true);
+    }
+
+    update_rosc_samples();
+
+    // Modify PRNG state with the two run-time entropy sources,
+    // hashed to reduce correlation with previous modifications.
+    rand_storage.s[0] ^= splitmix64(rosc_samples);
+    rand_storage.s[1] ^= splitmix64(time_us_64());
+
+    // Generate a 64-bit RN from the modified PRNG state.
+    // Note: This also "churns" the 128-bit state for next time.
+    return xoroshiro128ss();
+}
+
+
+void pico_initialise_rand(void *user_seed, bool force_ram_rehash) {
+
+    union unique_id_u_t unique_id;
+
+    // If supplied, use caller's seed to modify existing state.
+    // We further seed the state so the user seed is optional.
+    if (user_seed != NULL) {
+        struct prng_128_t *user_seed_128 = (struct prng_128_t *)user_seed;
+        rand_storage.s[0] ^= user_seed_128->s[0];
+        rand_storage.s[1] ^= user_seed_128->s[1];
+    }
+    // else - deliberately leave what is already in 'rand_storage'
+
+    // Note! The safety of the length assumption here is protected by a 'static_assert' above
+    pico_get_unique_board_id(&unique_id.board_id_native);
```

Comment:
This is less code (the unique_id_u_t union declaration isn't needed) and doesn't rely on a GCC's type punning extension for unions.
```pico_unique_board_id_t unique_id;
pico_get_unique_board_id(&unique_id);
uint64_t unique_id_int;
memcpy(&unique_id_int, &unique_id, sizeof(unique_id_int));
rand_storage.s[0] ^= splitmix64(unique_id_int);```